### PR TITLE
Mainmenu content browser: Fix URL escaping

### LINKED ISSUE #11576: Content Browser fails downloads/updates/installs/dependency checks when ContentDB username contains special characters
##### Minetest versions
5.3, 5.4.1, 5.5.0-dev (0c1e9603db32b4281974fdd8b8e3b505148be47e)

Running on Arch Linux with curl 7.78.0-1 installed

##### Summary
When attempting to do anything to a package from ContentDB that would involve web requests (ie. downloading, updating, checking package deps), and the package uploader's username contains a non-url-safe character (ie. my current username on CDB is `Hugues Ross`), the relevant curl requests fail with an error like:
> 2021-08-27 17:44:00: ERROR[Main]: https://content.minetest.net/api/packages/hugues ross/rpg16/dependencies/?only_hard=1&protocol_version=40&engine_version=5.5.0-dev not found (URL using bad/illegal format or missing URL) (response code 0)
2021-08-27 17:44:00: ERROR[AsyncWorker-0]: https://content.minetest.net/packages/Hugues Ross/rpg16/releases/9135/download/?reason=new not found (URL using bad/illegal format or missing URL) (response code 0)
2021-08-27 17:44:00: ERROR[AsyncWorker-0]: downloading "https://content.minetest.net/packages/Hugues Ross/rpg16/releases/9135/download/?reason=new" failed

I _suspect_ this relates to a curl update (more specifically, https://github.com/curl/curl/pull/7073), simply because this issue didn't exist in the past but seems to affect past versions. I'm looking into getting a name change as a backwards-compatible fix, but ideally we should properly fix our urls before passing them to curl.

##### Steps to reproduce
1. Open the content browser
2. Attempt to download or update any of [my packages](https://content.minetest.net/users/Hugues%20Ross/)

### PR DESCRIPTION
**Pull request by appgurueu**
Fixes #11576, ready for review. Do not squash. The diff is a bit large because I moved around the `table` utils (which was necessary) and added & documented the added string URI escaping utils. As this escapes spaces as `%20` and not `+`, https://github.com/minetest/contentdb/issues/335 doesn't need to be fixed right now.

### CODE REVIEW

**Comment by rubenwardy** [comment]
Curl has URL utilities

**Comment by appgurueu** [comment]
> Curl has URL utilities

Doesn't mean we have to use them.

**Comment by sfan5** [comment]
The only thing I'm wondering about is whether we should make these part of the supported API.
If we don't then using the curl functions would also be fine since they don't _need_ to work when it's disabled.

File: builtin/mainmenu/dlg_contentstore.lua [hunk 1]
```diff
@@ -578,11 +580,18 @@ function store.load()
 	store.aliases = {}
 
 	for _, package in pairs(store.packages_full) do
+		package.url = base_url .. "/packages/" ..
+			package.author:lower():encode_uri_component() ..
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_contentstore.lua:584]: author usernames in URLs should be considered case sensitive, so don't lower them

File: builtin/mainmenu/dlg_contentstore.lua [hunk 2]
```diff
@@ -578,11 +580,18 @@ function store.load()
...
+			"/" .. package.name ..
+			"/releases/" .. package.release .. "/download/"
+
 		local name_len = #package.name
 		if package.type == "game" and name_len > 5 and package.name:sub(name_len - 4) == "_game" then
-			package.id = package.author:lower() .. "/" .. package.name:sub(1, name_len - 5)
+			package.id = package.author:lower():encode_uri_component() ..
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_contentstore.lua:590]: id is used to track packages locally, and should never be used in a URL as case info is lost

File: builtin/mainmenu/dlg_contentstore.lua [hunk 3]
```diff
@@ -578,11 +580,18 @@ function store.load()
...
+				"/" .. package.name:sub(1, name_len - 5)
 		else
-			package.id = package.author:lower() .. "/" .. package.name
+			package.id = package.author:lower():encode_uri_component() ..
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_contentstore.lua:593]: id is used to track packages locally, and should never be used in a URL as case info is lost

File: builtin/mainmenu/dlg_contentstore.lua [hunk 4]
```diff
@@ -606,7 +615,7 @@ function store.update_paths()
 	pkgmgr.refresh_globals()
 	for _, mod in pairs(pkgmgr.global_mods:get_list()) do
 		if mod.author and mod.release > 0 then
-			local id = mod.author:lower() .. "/" .. mod.name
+			local id = mod.author:lower():encode_uri_component() .. "/" .. mod.name
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_contentstore.lua:618]: id is used to track packages locally, and should never be used in a URL as case info is lost

File: builtin/mainmenu/dlg_contentstore.lua [hunk 5]
```diff
@@ -615,15 +624,15 @@ function store.update_paths()
 	pkgmgr.update_gamelist()
 	for _, game in pairs(pkgmgr.games) do
 		if game.author ~= "" and game.release > 0 then
-			local id = game.author:lower() .. "/" .. game.id
+			local id = game.author:lower():encode_uri_component() .. "/" .. game.id
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_contentstore.lua:627]: id is used to track packages locally, and should never be used in a URL as case info is lost

File: builtin/mainmenu/dlg_contentstore.lua [hunk 6]
```diff
@@ -615,15 +624,15 @@ function store.update_paths()
...
 			game_hash[store.aliases[id] or id] = game
 		end
 	end
 
 	local txp_hash = {}
 	for _, txp in pairs(pkgmgr.get_texture_packs()) do
 		if txp.author and txp.release > 0 then
-			local id = txp.author:lower() .. "/" .. txp.name
+			local id = txp.author:lower():encode_uri_component() .. "/" .. txp.name
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_contentstore.lua:635]: id is used to track packages locally, and should never be used in a URL as case info is lost

File: doc/client_lua_api.txt [hunk 7]
```diff
@@ -607,6 +607,11 @@ Helper functions
     * e.g. `string:split("a,b", ",") == {"a","b"}`
 * `string:trim()`
     * e.g. `string.trim("\n \t\tfoo bar\t ") == "foo bar"`
+* `string:encode_uri()`: Equivalent to the JavaScript one, see [`encodeURI`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI)
+	* Allows reserved & unescaped characters
```
Comment (rubenwardy) [inline doc/client_lua_api.txt:611]: I wonder if you should use the examples from MDN to make this clear

**Review by rubenwardy — COMMENTED** [review]
You haven't updated the `menu_lua_api.txt` doc

**Comment by rubenwardy** [comment]
I think it's perfectly fine to not have these util functions available when cURL isn't in use. The reason I suggested using the cURL utils is because they are well tested, but I'm not too bothered about that as they aren't too complicated

**Comment by appgurueu** [comment]
> You haven't updated the `menu_lua_api.txt` doc

Do I have to? These utils aren't only available to the main menu.

**Comment by sfan5** [comment]
The function should at least have an entry there, even without documentation.

**Comment by appgurueu** [comment]
> The function should at least be mentioned there, even if without documentation.

That's a redundant mess. I'd have to copy it to the Client Lua API as well. See #11577.

**Comment by rubenwardy** [comment]
You already have copied it to the client API reference, it just needs to be copied to the main menu one as well

**Comment by appgurueu** [comment]
> You already have copied it to the client API reference, it just needs to be copied to the main menu one as well

I have? In that case, I will happily oblige for now, until #11577 gets done.

File: doc/client_lua_api.txt (hunk 7 above)
Comment (appgurueu) [reply doc/client_lua_api.txt:611]: Is that really necessary? We have zlib compression, base64 etc. without any examples as it's clearly specified externally already.

File: builtin/mainmenu/dlg_contentstore.lua (hunk 1 above)
Comment (appgurueu) [reply builtin/mainmenu/dlg_contentstore.lua:599]: I just preserved the `:lower()` from before, but I can remove it if you wish.

Comment (rubenwardy) [reply builtin/mainmenu/dlg_contentstore.lua:599]: The `:lower()` is only on `.id` fields, which is correct

This `.url` isn't used at all it looks like

**Comment by rubenwardy** [comment]
Replaced by #11814

**Pull request closed without merging**